ddprof-lib/src/main/cpp/profiler.cpp

```diff
@@ -419,27 +411,20 @@ int Profiler::getJavaTraceAsync(void* ucontext, ASGCT_CallFrame* frames, int max
         _thread_max_state         = 12  // maximum thread state+1 - used for statistics allocation
     };
      */
-    // avoid unwinding during deoptimization
-    if (vm_thread->osThreadState() == ThreadState::RUNNABLE && (state == 10 || state == 11)) {
-        return 0;
-    }
     bool in_java = (state == 8 || state == 9);
     if (in_java && java_ctx->sp != 0) {
         // skip ahead to the Java frames before calling AGCT
         frame.restore((uintptr_t)java_ctx->pc, java_ctx->sp, java_ctx->fp);
+    } else if (state != 0 && vm_thread->lastJavaSP() == 0) {
+        // we haven't found the top Java frame ourselves, and the lastJavaSP wasn't recorded either
+        Counters::increment(AGCT_NATIVE_NO_JAVA_CONTEXT);
+        return 0;
```

Comment:
Just for clarification - when `lastJavaSP()` is not zero it will be used later to jump to the Java frame, right?

Reply:
Yes, AsyncGetCallTrace will use it in any non-Java thread state